# Table of content

- <a href='#intro'>Introduction</a>
- <a href='#data-wrangling'>Data Wrangling</a>
  - <a href='#gdp-ppp-dollars'>GDP in PPP dollars</a>
  - <a href='#gdp-current-dollars'>GDP in current USD</a>
  - <a href='#total-population'>Total population</a>
  - <a href='#fuel-exports'>Fuel exports</a>
  - <a href='#democracy-index'>Democracy index</a>
  - <a href='#join-data'>Join the data sets</a>

<a id='intro'></a>
# Introduction

In this project we investigate data sets on
GDP per capita,
democracy indices,
total population,
and fuel exports,
available
from [www.gapminder.org](https://www.gapminder.org/)
and [World Bank Open Data](https://data.worldbank.org/).
The data was available under the [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
from both sources, see
[free material from Gapminder](https://www.gapminder.org/free-material/)
and
[Summary of the Terms of Use for Datasets Listed in The World Bank Data Catalog](https://data.worldbank.org/summary-terms-of-use).


<a id='data-wrangling'></a>
# Data Wrangling


In [1]:
from common import *

import pandas as pd

In [2]:
%load_ext autoreload
%autoreload

In [3]:
print(f"We are going to analyze data sets for the year {the_year}.")

We are going to analyze data sets for the year 2018.


### Helper functions

First we define helper functions for data wrangling.

In [4]:
def read_csv(filepath, usecols = None, col_names = None, indexcol = None, header = 0):
    """
    Convenience function to read selected columns from a CSV and assign them custom names.

    :param filepath: Path to the CSV file to be read
    :param usecols: Original columns to be used. If None then all columns will be read
    :param col_names: Custom column names to be set after reading
    :param indexcol: Column to be set as index column
    :param header: Row number to be used as header containing the original column names
    :return:
    """
    df = pd.read_csv(filepath, usecols=usecols, header=header)
    if usecols is not None:
        df.columns = col_names
    if indexcol is not None:
        df.set_index(indexcol, inplace=True)
    return df

<a id='gdp-ppp-dollars'></a>
## GDP data, in PPP dollars

We investigate the data set "GD001: GDP per capita, constant PPP dollars",
available [here](https://www.gapminder.org/data/documentation/gd001/) from Gapminder.
We consider the data on GDP per capita in [PPP (Purchasing Power Parity)](https://en.wikipedia.org/wiki/Purchasing_power_parity)
dollars for the year 2018,
compiled by [Gapminder](https://www.gapminder.org/)
from the data provided by the World Bank [here](https://data.worldbank.org/indicator/NY.GDP.PCAP.PP.KD).

We will use pre-defined column names.

In [5]:
print_declared_class_attributes(ColGdp)

class ColGdp:
	country_code = "Country code"
	country = "Country"
	year = "Year"
	gdp_per_capita = "GDP per capita"
	gdp_per_capita_ppp = "GDP per capita (PPP dollars)"
	above_avg_gdp_per_capita = "Above-average GDP per capita"
	above_avg_gdp_per_capita_ppp = "Above-average GDP per capita (PPP dollars)"
	gdp_level = "GDP level"
	gdp_total = "GDP total"
	gdp_total_ppp = "GDP total (PPP dollars)"


In [6]:
df_gdp_ppp = read_csv('../resources/project/GM-GDP-per-capita_Dataset-v26_data-for-contries-by-year.csv',
    header=0,
    usecols=["geo","name","time","Income per person"],
    col_names=[ColGdp.country_code, ColGdp.country, ColGdp.year, ColGdp.gdp_per_capita_ppp],
    indexcol=ColGdp.country_code)
df_gdp_ppp = df_gdp_ppp[df_gdp_ppp[ColGdp.year] == the_year]
df_gdp_ppp.head()

Country  Year  GDP per capita (PPP dollars)
Country code                                                 
afg           Afghanistan  2018                          1735
alb               Albania  2018                         12306
dza               Algeria  2018                         13886
and               Andorra  2018                         51494
ago                Angola  2018                          5725

Which countries are there in the dataset?

In [7]:
print(f"There are {df_gdp_ppp[ColGdp.country].nunique()} countries in the data set:\n"
      f"{df_gdp_ppp[ColGdp.country].unique()}")

There are 195 countries in the data set:
['Afghanistan' 'Albania' 'Algeria' 'Andorra' 'Angola'
 'Antigua and Barbuda' 'Argentina' 'Armenia' 'Australia' 'Austria'
 'Azerbaijan' 'Bahamas' 'Bahrain' 'Bangladesh' 'Barbados' 'Belarus'
 'Belgium' 'Belize' 'Benin' 'Bhutan' 'Bolivia' 'Bosnia and Herzegovina'
 'Botswana' 'Brazil' 'Brunei' 'Bulgaria' 'Burkina Faso' 'Burundi'
 'Cambodia' 'Cameroon' 'Canada' 'Cape Verde' 'Central African Republic'
 'Chad' 'Chile' 'China' 'Colombia' 'Comoros' 'Congo, Dem. Rep.'
 'Congo, Rep.' 'Costa Rica' "Cote d'Ivoire" 'Croatia' 'Cuba' 'Cyprus'
 'Czech Republic' 'Denmark' 'Djibouti' 'Dominica' 'Dominican Republic'
 'Ecuador' 'Egypt' 'El Salvador' 'Equatorial Guinea' 'Eritrea' 'Estonia'
 'Ethiopia' 'Fiji' 'Finland' 'France' 'Gabon' 'Gambia' 'Georgia' 'Germany'
 'Ghana' 'Greece' 'Grenada' 'Guatemala' 'Guinea' 'Guinea-Bissau' 'Guyana'
 'Haiti' 'Honduras' 'Hong Kong, China' 'Hungary' 'Iceland' 'India'
 'Indonesia' 'Iran' 'Iraq' 'Ireland' 'Israel' 'Italy' 'Jamaica' 'J

In [8]:
df_gdp_ppp.info()

<class 'pandas.core.frame.DataFrame'>
Index: 195 entries, afg to zwe
Data columns (total 3 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   Country                       195 non-null    object
 1   Year                          195 non-null    int64 
 2   GDP per capita (PPP dollars)  195 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 6.1+ KB


There are 195 rows and no missing values in the data set.

In [9]:
df_gdp_ppp.duplicated().sum()

0

There are no duplicated rows in the data set.

In [10]:
df_gdp_ppp[ColGdp.country].duplicated().sum()

0

There are also no duplicate country values:
each country has exactly one entry for the considered year 2018.

<a id='gdp-current-dollars'></a>
## GDP data, in current USD

We investigate the data set "GDP per capita (current US$)",
available [here](https://data.worldbank.org/indicator/NY.GDP.PCAP.CD) from the World Bank.
We consider the data on GDP per capita in current US dollars for the year 2018.

In [11]:
df_gdp_curr = read_csv('../resources/project/API_NY.GDP.PCAP.CD_DS2_en_csv_v2_2055804_WorldBank_GDP_per_capita_current_USD.csv',
    header=2,
    usecols=["Country Name", "Country Code", "2018"],
    col_names=[ColGdp.country, ColGdp.country_code, ColGdp.gdp_per_capita],
    indexcol=ColGdp.country_code) # the actual header row in the dataset is preceded by some other information
df_gdp_curr.head()

Country  GDP per capita
Country code                             
ABW                 Aruba             NaN
AFG           Afghanistan      493.750418
AGO                Angola     3289.646664
ALB               Albania     5284.380184
AND               Andorra    41793.055258

Convert the country codes to lower case so that it matches the country codes in `df_gdp_ppp`.

In [12]:
df_gdp_curr.index = df_gdp_curr.index.str.lower()
df_gdp_curr.head()

Country  GDP per capita
Country code                             
abw                 Aruba             NaN
afg           Afghanistan      493.750418
ago                Angola     3289.646664
alb               Albania     5284.380184
and               Andorra    41793.055258

In [13]:
df_gdp_curr.info()

<class 'pandas.core.frame.DataFrame'>
Index: 264 entries, abw to zwe
Data columns (total 2 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country         264 non-null    object 
 1   GDP per capita  249 non-null    float64
dtypes: float64(1), object(1)
memory usage: 6.2+ KB


There are some missing values. We will deal with them after joining the two GDP data frames in the next subsection.

## Join the GDP data frames

We join the `df_dgp_curr` data frame containing GDP per capita in actual USD,
and the `df_gdp_ppp` data frame containing GDP in PPP dollars.

In [14]:
df_gdp = df_gdp_ppp.join(df_gdp_curr[[ColGdp.gdp_per_capita]])
df_gdp.head()

Country  Year  GDP per capita (PPP dollars)  GDP per capita
Country code                                                                 
afg           Afghanistan  2018                          1735      493.750418
alb               Albania  2018                         12306     5284.380184
dza               Algeria  2018                         13886     4153.733978
and               Andorra  2018                         51494    41793.055258
ago                Angola  2018                          5725     3289.646664

In [15]:
df_gdp.describe()

Year  GDP per capita (PPP dollars)  GDP per capita
count   195.0                    195.000000      188.000000
mean   2018.0                  18786.097436    15558.870464
std       0.0                  19779.346476    23448.392538
min    2018.0                    629.000000      271.752044
25%    2018.0                   3779.500000     2158.678618
50%    2018.0                  12143.000000     6117.407428
75%    2018.0                  28209.500000    17394.775949
max    2018.0                 112532.000000   185829.017960

There are a few countries where the GDP per capita is missing.

In [16]:
df_gdp[df_gdp[ColGdp.gdp_per_capita].isnull()]

Country  Year  GDP per capita (PPP dollars)  GDP per capita
Country code                                                                 
eri               Eritrea  2018                          1374             NaN
prk           North Korea  2018                          1716             NaN
som               Somalia  2018                           629             NaN
ssd           South Sudan  2018                          1772             NaN
syr                 Syria  2018                          2900             NaN
twn                Taiwan  2018                         43273             NaN
ven             Venezuela  2018                         12468             NaN

We will substitute the missing values by the available values in PPP dollars:
although this is not very accurate, it should be sufficient for the envisioned analysis.
We will only make more precise manual adjustments for two of these countries.

In [17]:
df_gdp[ColGdp.gdp_per_capita].fillna(df_gdp[ColGdp.gdp_per_capita_ppp], inplace=True)
df_gdp.isnull().sum()

Country                         0
Year                            0
GDP per capita (PPP dollars)    0
GDP per capita                  0
dtype: int64

Now there are no more missing values.

#### Manual adjustments for selected countries

We separately adjust the data points for [Venezuela](https://en.wikipedia.org/wiki/Venezuela)
and [Taiwan](https://en.wikipedia.org/wiki/Taiwan).

[Venezuela](https://en.wikipedia.org/wiki/Venezuela)
is a country with large [oil and natural gas reserves](https://en.wikipedia.org/wiki/Venezuela#Petroleum_and_other_resources)
that experienced a significant decline in both
[GDP](https://countryeconomy.com/gdp/venezuela#:~:text=The%20GDP%20per%20capita%20of,2017%2C%20when%20it%20was%20%244%2C894.)
and [democracy index](https://en.wikipedia.org/wiki/Democracy_Index#By_country)
since 2015.
The GDP per capita fell from ca. 10,000 USD in 2015 to less than 4,000 USD in 2018.
We will approximate the missing GDP value with the latter figure.

In [18]:
df_gdp.at['ven', ColGdp.gdp_per_capita] = 4000
df_gdp[ColGdp.gdp_per_capita]['ven']

4000.0

[Taiwan](https://en.wikipedia.org/wiki/Taiwan), officially the Republic of China (ROC),
is a developed democratic country in East Asia.
It's political status is disputed because it's considered by China as one if its provinces.
Only few other countries maintain official diplomatic relations with it.
Yet, many contries, including China, cooperate economically.
Taiwans high-tech industry plays an important role for economies worldwide.
As an example, its semiconductor contract manufacturers have
[over 50% of world market share](https://www.cnbc.com/2021/03/16/2-charts-show-how-much-the-world-depends-on-taiwan-for-semiconductors.html).

Taiwans disputed political status could also be a reason why it's missing in some data sets,
esp. the ones from the World Bank,
so that we have to separately collect it from other sources,
e.g., [Wikipedia](https://en.wikipedia.org/wiki/Economy_of_Taiwan).
Based on various sources on the Internet, we will approximate its GDP per capita for 2018 by 25,000 in current USD.

In [19]:
df_gdp.at['twn', ColGdp.gdp_per_capita] = 25000
df_gdp[ColGdp.gdp_per_capita]['twn']

25000.0

<a id='total-population'></a>
## The total population data set

We investigate the data set "Total population",
available [here](https://data.worldbank.org/indicator/SP.POP.TOTL) from the World Bank.
We consider the data on total population for the year 2018.

In [20]:
print_declared_class_attributes(ColPop)

class ColPop:
	country_code = "Country code"
	country = "Country"
	population = "Population"


In [21]:
df_pop = read_csv('../resources/project/API_SP.POP.TOTL_DS2_en_csv_v2_2163507_WorldBank_total_population.csv',
    header=2,
    usecols=["Country Name", "Country Code", f"{the_year}"],
    col_names=[ColPop.country, ColPop.country_code, ColPop.population],
    indexcol=ColPop.country_code)
df_pop.head()

Country  Population
Country code                         
ABW                 Aruba    105845.0
AFG           Afghanistan  37172386.0
AGO                Angola  30809762.0
ALB               Albania   2866376.0
AND               Andorra     77006.0

Convert the country codes to lower case so that it matches the country codes in the other data frames.

In [22]:
df_pop.index = df_pop.index.str.lower()
df_pop.head()

Country  Population
Country code                         
abw                 Aruba    105845.0
afg           Afghanistan  37172386.0
ago                Angola  30809762.0
alb               Albania   2866376.0
and               Andorra     77006.0

In [23]:
df_pop.info()

<class 'pandas.core.frame.DataFrame'>
Index: 264 entries, abw to zwe
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Country     264 non-null    object 
 1   Population  262 non-null    float64
dtypes: float64(1), object(1)
memory usage: 6.2+ KB


In [24]:
df_pop[df_pop[ColPop.population].isnull()]

Country  Population
Country code                            
eri                  Eritrea         NaN
inx           Not classified         NaN

There are two rows with missing population figures. We will omit these rows for our analysis.

In [25]:
df_pop.dropna(inplace=True)

In [26]:
df_pop.sort_values(by=ColPop.population, ascending=False)

Country    Population
Country code                                      
wld                            World  7.591945e+09
ibt                 IDA & IBRD total  6.410739e+09
lmy              Low & middle income  6.361430e+09
mic                    Middle income  5.710102e+09
ibd                        IBRD only  4.777257e+09
...                              ...           ...
gib                        Gibraltar  3.371800e+04
vgb           British Virgin Islands  2.980200e+04
plw                            Palau  1.790700e+04
nru                            Nauru  1.270400e+04
tuv                           Tuvalu  1.150800e+04

[262 rows x 2 columns]

It turns out that the data set contains more values, than just the population by country.
For example, it has a row for the total world population and also rows for population by region and income.
This data would only disturb our analysis, so we filter the data set to contain only countries
that are also present in the other data sets.

In [27]:
df_pop = df_pop.filter(df_gdp.index, axis='index')
df_pop.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, afg to zwe
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Country     193 non-null    object 
 1   Population  193 non-null    float64
dtypes: float64(1), object(1)
memory usage: 4.5+ KB


The `population` column should contain integers, but was imported as `float64`,
probably because of the rows with missing values.
We will convert the column to `int64`.

In [28]:
df_pop[ColPop.population] = df_pop[ColPop.population].astype('int64')
df_pop.head()

Country  Population
Country code                         
afg           Afghanistan    37172386
alb               Albania     2866376
dza               Algeria    42228429
and               Andorra       77006
ago                Angola    30809762

In [29]:
df_pop.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, afg to zwe
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Country     193 non-null    object
 1   Population  193 non-null    int64 
dtypes: int64(1), object(1)
memory usage: 4.5+ KB


As for GDP, the data for Taiwan is missing. We approximate it by 23 million.

In [30]:
df_pop.at['twn', ColPop.population] = 23000000
df_pop[ColPop.population]['twn']

23000000.0

Now the population data set has been cleaned for our intended analysis.

<a id='fuel-exports'></a>
## The fuel exports data set

We investigate the data set "Fuel exports (% of merchandise exports)",
available [here](https://data.worldbank.org/indicator/TX.VAL.FUEL.ZS.UN) from the World Bank.
We consider the data on fuel exports for the year 2018.

In [31]:
print_declared_class_attributes(ColFuel)

class ColFuel:
	country_code = "Country code"
	country = "Country"
	fuel_exports = "Fuel exports (% of merchandise exports)"


We expect some missing values for 2018,
so we also import the data for 2 previous years in order to fill the gaps.

In [32]:
fuel_exports_2_years_ago = "Fuel exports 2 years ago"
fuel_exports_1_year_ago = "Fuel exports 1 year ago"
df_fuel = read_csv('../resources/project/API_TX.VAL.FUEL.ZS.UN_DS2_en_csv_v2_2167156_WorldBank_fuel_exports.csv',
    header=2,
    usecols=["Country Name", "Country Code", f"{the_year - 2}", f"{the_year - 1}", f"{the_year}"],
    col_names=[ColFuel.country, ColFuel.country_code, fuel_exports_2_years_ago, fuel_exports_1_year_ago, ColFuel.fuel_exports],
    indexcol=ColFuel.country_code)
df_fuel.head()

Country  Fuel exports 2 years ago  Fuel exports 1 year ago  \
Country code                                                                   
ABW                 Aruba                  0.104982                 0.114804   
AFG           Afghanistan                       NaN                      NaN   
AGO                Angola                 92.701056                94.691252   
ALB               Albania                 11.175063                 0.824033   
AND               Andorra                  0.010607                 0.008582   

              Fuel exports (% of merchandise exports)  
Country code                                           
ABW                                          0.059140  
AFG                                         10.213180  
AGO                                         92.417810  
ALB                                          1.662666  
AND                                          0.013639

Convert the country codes to lower case so that it matches the country codes in the other data frames.

In [33]:
df_fuel.index = df_fuel.index.str.lower()
df_fuel.head()

Country  Fuel exports 2 years ago  Fuel exports 1 year ago  \
Country code                                                                   
abw                 Aruba                  0.104982                 0.114804   
afg           Afghanistan                       NaN                      NaN   
ago                Angola                 92.701056                94.691252   
alb               Albania                 11.175063                 0.824033   
and               Andorra                  0.010607                 0.008582   

              Fuel exports (% of merchandise exports)  
Country code                                           
abw                                          0.059140  
afg                                         10.213180  
ago                                         92.417810  
alb                                          1.662666  
and                                          0.013639

In [34]:
df_fuel[df_fuel[ColFuel.fuel_exports].isnull()]

Country  Fuel exports 2 years ago  \
Country code                                                     
asm                   American Samoa                       NaN   
bgd                       Bangladesh                       NaN   
btn                           Bhutan                       NaN   
chi                  Channel Islands                       NaN   
cmr                         Cameroon                  7.692626   
...                              ...                       ...   
vgb           British Virgin Islands                       NaN   
vir            Virgin Islands (U.S.)                       NaN   
vut                          Vanuatu                       NaN   
xkx                           Kosovo                       NaN   
yem                      Yemen, Rep.                       NaN   

              Fuel exports 1 year ago  Fuel exports (% of merchandise exports)  
Country code                                                                    
asm                               NaN                                      NaN  
bgd                               NaN                                      NaN  
btn                               NaN                                      NaN  
chi                               NaN                                      NaN  
cmr                         43.449411                                      NaN  
...                               ...                                      ...  
vgb                               NaN                                      NaN  
vir                               NaN                                      NaN  
vut                               NaN                                      NaN  
xkx                               NaN                                      NaN  
yem                               NaN                                      NaN  

[79 rows x 4 columns]

There are a lot of missing values.
We fill try to fill them from the previous years.

In [35]:
df_fuel.loc[:, fuel_exports_2_years_ago:] = df_fuel.loc[:, fuel_exports_2_years_ago:].fillna(method='ffill', axis='columns')
df_fuel[ColFuel.fuel_exports].isnull().sum()

62

There are still a lot of missing values: we now assume that those are no longer relevant for our analysis
and fill them with 0.

In [36]:
df_fuel.fillna(0, inplace=True)

We drop the columns for previous years because we only needed them to fill missing values.

In [37]:
df_fuel.drop(columns=[fuel_exports_2_years_ago, fuel_exports_1_year_ago], inplace=True)
df_fuel.head()

Country  Fuel exports (% of merchandise exports)
Country code                                                      
abw                 Aruba                                 0.059140
afg           Afghanistan                                10.213180
ago                Angola                                92.417810
alb               Albania                                 1.662666
and               Andorra                                 0.013639

In [38]:
df_fuel.describe()

Fuel exports (% of merchandise exports)
count                               264.000000
mean                                 14.150729
std                                  23.942621
min                                   0.000000
25%                                   0.000008
50%                                   2.114630
75%                                  15.619748
max                                  99.986485

In [39]:
df_fuel.loc[['wld', 'mea', 'oed']]

Country  \
Country code                               
wld                                World   
mea           Middle East & North Africa   
oed                         OECD members   

              Fuel exports (% of merchandise exports)  
Country code                                           
wld                                         14.251782  
mea                                         67.595873  
oed                                          8.382116

This data set, similarly to the population data set, contains more values than just the fuel exports by country.
For example, it has a row for the world and for some world regions.
This data would only disturb our analysis, so we filter the data set to contain only countries
that are also present in the other data sets.

In [40]:
df_fuel = df_fuel.filter(df_gdp.index, axis='index')
df_fuel.info()

<class 'pandas.core.frame.DataFrame'>
Index: 194 entries, afg to zwe
Data columns (total 2 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   Country                                  194 non-null    object 
 1   Fuel exports (% of merchandise exports)  194 non-null    float64
dtypes: float64(1), object(1)
memory usage: 4.5+ KB


Now the fuel export data set has been cleaned for our intended analysis.

<a id='democracy-index'></a>
## The democracy index data set

We investigate the data set "Democracy indices" published by the
[The Economist Intelligence Unit](https://www.eiu.com/),
available [here](https://data.worldbank.org/indicator/NY.GDP.PCAP.CD) from the World Bank.

We consider the overall democracy index scores for the year 2018,
calculated from five category indices measuring
electoral process and pluralism, functioning of government, political participation, political culture, and civil liberties.
Further freely available explanation on the democracy index can be found in the [Wikipedia article](https://en.wikipedia.org/wiki/Democracy_Index).

In [41]:
print_declared_class_attributes(ColDem)

class ColDem:
	country_code = "Country code"
	country = "Country"
	year = "Year"
	democracy_index = "Democracy index"
	regime_type = "Regime type"


In [42]:
df_dem = read_csv('../resources/project/EIU-Democracy-Indices_Dataset-v3_data-for-countries-by-year.csv',
    usecols=["geo", "name", "time", "Democracy index (EIU)"],
    col_names=[ColDem.country_code, ColDem.country, ColDem.year, ColDem.democracy_index],
    indexcol=ColDem.country_code)
df_dem = df_dem[df_dem[ColDem.year] == the_year].copy()
df_dem.head()

Country  Year  Democracy index
Country code                                    
afg           Afghanistan  2018             29.7
alb               Albania  2018             59.8
dza               Algeria  2018             35.0
ago                Angola  2018             36.2
arg             Argentina  2018             70.2

In [43]:
df_dem.info()

<class 'pandas.core.frame.DataFrame'>
Index: 167 entries, afg to zwe
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country          167 non-null    object 
 1   Year             167 non-null    int64  
 2   Democracy index  167 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 5.2+ KB


There are 167 rows and no missing values in the data set.

We now calcluate the regime types. The EUI differentiates between
full democracy, flawed democracy, hybrid regime, and authoritarian regime,
cf. [EUI's democracy index visualization](https://infographics.economist.com/2018/DemocracyIndex/)
or the [Wikipedia article on Democracy Index](https://en.wikipedia.org/wiki/Democracy_Index#By_regime_type).
We simplify this differentiation to the 3 categories Democracy, Hybrid, and Authoritarian:

| Type of Regime | Democracy Index |
|:---------------|:----------------|
| Democracy | 6 - 10 |
| Hybrid | 4 - 6 |
| Authoritarian | 0 - 4 |

In [44]:
print_declared_class_attributes(RegimeType)

class RegimeType:
	democracy = "Democracy"
	hybrid = "Hybrid"
	authoritarian = "Authoritarian"
	values = "['Democracy', 'Hybrid', 'Authoritarian']"


In [45]:
regime_types = [RegimeType.authoritarian, RegimeType.hybrid, RegimeType.democracy]
regime_type_edges = [0, 40, 60, 100]
df_dem[ColDem.regime_type] = pd.cut(df_dem[ColDem.democracy_index], bins=regime_type_edges, labels=regime_types)
df_dem.head()

Country  Year  Democracy index    Regime type
Country code                                                   
afg           Afghanistan  2018             29.7  Authoritarian
alb               Albania  2018             59.8         Hybrid
dza               Algeria  2018             35.0  Authoritarian
ago                Angola  2018             36.2  Authoritarian
arg             Argentina  2018             70.2      Democracy

### Assess missing data in democracy data frame

In [46]:
print(f"Number of countries in GDP data frame: {len(df_gdp)}")
print(f"Number of countries in Democracy data frame: {len(df_dem)}")
print(f"There are {df_gdp.index.difference(df_dem.index).size} countries with GDP data but no democracy index data")

Number of countries in GDP data frame: 195
Number of countries in Democracy data frame: 167
There are 28 countries with GDP data but no democracy index data


The democracy index data is missing for 28 countries.
We want to check how much of GDP and population they represent to assess
if we can omit them from our analysis.

In [47]:
df_dem_missing = df_gdp.loc[df_gdp.index.difference(df_dem.index)]\
    .join(df_pop[ColPop.population])
df_dem_missing

Country  Year  \
Country code                                         
and                                  Andorra  2018   
atg                      Antigua and Barbuda  2018   
bhs                                  Bahamas  2018   
blz                                   Belize  2018   
brb                                 Barbados  2018   
brn                                   Brunei  2018   
dma                                 Dominica  2018   
fsm                    Micronesia, Fed. Sts.  2018   
grd                                  Grenada  2018   
kir                                 Kiribati  2018   
kna                      St. Kitts and Nevis  2018   
lca                                St. Lucia  2018   
mco                                   Monaco  2018   
mdv                                 Maldives  2018   
mhl                         Marshall Islands  2018   
nru                                    Nauru  2018   
plw                                    Palau  2018   
slb                          Solomon Islands  2018   
smr                               San Marino  2018   
som                                  Somalia  2018   
ssd                              South Sudan  2018   
stp                    Sao Tome and Principe  2018   
syc                               Seychelles  2018   
ton                                    Tonga  2018   
tuv                                   Tuvalu  2018   
vct           St. Vincent and the Grenadines  2018   
vut                                  Vanuatu  2018   
wsm                                    Samoa  2018   

              GDP per capita (PPP dollars)  GDP per capita  Population  
Country code                                                            
and                                  51494    41793.055258     77006.0  
atg                                  23768    16672.744240     96286.0  
bhs                                  28833    33767.503371    385640.0  
blz                                   7810     4884.734162    383071.0  
brb                                  16698    17745.193465    286641.0  
brn                                  71802    31628.328791    428962.0  
dma                                   9467     7693.878870     71625.0  
fsm                                   3196     3568.291016    112640.0  
grd                                  13970    10485.907157    111454.0  
kir                                   2035     1698.256286    115847.0  
kna                                  28295    19276.521267     52438.0  
lca                                  12344    11357.889010    181889.0  
mco                                  60712   185829.017960     38682.0  
mdv                                  13611    10276.932815    515696.0  
mhl                                   3599     3788.163594     58413.0  
nru                                  13373     9762.389302     12704.0  
plw                                  17202    15660.674597     17907.0  
slb                                   2142     2441.517393    652858.0  
smr                                  57647    48995.142002     33785.0  
som                                    629      629.000000  15008154.0  
ssd                                   1772     1772.000000  10975920.0  
stp                                   3033     1953.588290    211028.0  
syc                                  27114    16390.824908     96762.0  
ton                                   5696     4740.792149    103197.0  
tuv                                   3593     3700.744263     11508.0  
vct                                  10940     7361.400962    110210.0  
vut                                   2846     3125.257989    292680.0  
wsm                                   6089     4188.528346    196130.0

It looks like almost all of those countries are quite small, most of them having a population of less than 1M,
and some even less than 100.000.

We now calculate the share of their population and GDP relatively to all countries in the GDP data set.

In [48]:
pop_total = df_pop[ColPop.population].sum()
pop_total_missing_in_df_dem = df_dem_missing[ColPop.population].sum()

gdp_total = (df_gdp[ColGdp.gdp_per_capita] * df_gdp.join(df_pop[ColPop.population])[ColPop.population]).sum()
gdp_total_missing_in_df_dem = (df_dem_missing[ColGdp.gdp_per_capita] * df_dem_missing[ColPop.population]).sum()

The following data frame shows the population, the GDP, and the percentage of population and GDP
in countries with missing democracy data in relation to all countries in the GDP data set:

In [49]:
pd.DataFrame(
    data=[[pop_total, gdp_total],
          [pop_total_missing_in_df_dem, gdp_total_missing_in_df_dem],
          [f"{pop_total_missing_in_df_dem / pop_total * 100:.2f}%", f"{gdp_total_missing_in_df_dem / gdp_total * 100:.2f}%"]],
    index=["All countries", "Countries without democracy index data", "Percentage for countries without democracy index data "],
    columns=["Total Population", "Total GDP"])

Total Population  \
All countries                                           7.58038e+09   
Countries without democracy index data                  3.06391e+07   
Percentage for countries without democracy inde...            0.40%   

                                                      Total GDP  
All countries                                       8.57333e+13  
Countries without democracy index data               9.4096e+10  
Percentage for countries without democracy inde...        0.11%

The countries with missing democracy data represent
around 0.4% of world population
and around 0.1% of world GDP.
Hence, leaving them out would not significantly impact the outcome of our analysis.

<a id='join-data'></a>
## Join all data sets

Join the GDP per capita in the `df_gdp` data frame
with the `df_pop` data frame containing population values,
the `df_fuel` data frame containing fuel export values,
and the `df_dem` data frame containing democracy index values.

We will use inner join to consider only data on countries,
for which every data frame has values available.
Only for fuel exports, we will use left join and fill missing values with 0.

In [50]:
df = df_gdp.join(df_pop[ColPop.population], how="inner")\
    .join(df_dem[[ColDem.democracy_index, ColDem.regime_type]], how="inner")\
    .join(df_fuel[ColFuel.fuel_exports], how="left")\

df[ColFuel.fuel_exports].fillna(0, inplace=True)
df.head()

Country  Year  GDP per capita (PPP dollars)  GDP per capita  \
Country code                                                                    
afg           Afghanistan  2018                          1735      493.750418   
alb               Albania  2018                         12306     5284.380184   
dza               Algeria  2018                         13886     4153.733978   
ago                Angola  2018                          5725     3289.646664   
arg             Argentina  2018                         18282    11633.498009   

              Population  Democracy index    Regime type  \
Country code                                               
afg           37172386.0             29.7  Authoritarian   
alb            2866376.0             59.8         Hybrid   
dza           42228429.0             35.0  Authoritarian   
ago           30809762.0             36.2  Authoritarian   
arg           44494502.0             70.2      Democracy   

              Fuel exports (% of merchandise exports)  
Country code                                           
afg                                         10.213180  
alb                                          1.662666  
dza                                         96.111562  
ago                                         92.417810  
arg                                          4.119375

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 166 entries, afg to zwe
Data columns (total 8 columns):
 #   Column                                   Non-Null Count  Dtype   
---  ------                                   --------------  -----   
 0   Country                                  166 non-null    object  
 1   Year                                     166 non-null    int64   
 2   GDP per capita (PPP dollars)             166 non-null    int64   
 3   GDP per capita                           166 non-null    float64 
 4   Population                               166 non-null    float64 
 5   Democracy index                          166 non-null    float64 
 6   Regime type                              166 non-null    category
 7   Fuel exports (% of merchandise exports)  166 non-null    float64 
dtypes: category(1), float64(4), int64(2), object(1)
memory usage: 15.6+ KB


We now have a data frame of 166 countries
containing data on
GDP per capita (in PPP dollars and in current USD),
population,
democracy level,
and fuel exports
for the year 2018.

### Derive additional columns

We have already added the `regime_type` columns derived from democracy index.
We now add columns with total GDP per country, calculated from GDP per capita and population size.

In [52]:
df[ColGdp.gdp_total] = df[ColGdp.gdp_per_capita] * df[ColPop.population]
df[ColGdp.gdp_total_ppp] = df[ColGdp.gdp_per_capita_ppp] * df[ColPop.population]
df.head()

Country  Year  GDP per capita (PPP dollars)  GDP per capita  \
Country code                                                                    
afg           Afghanistan  2018                          1735      493.750418   
alb               Albania  2018                         12306     5284.380184   
dza               Algeria  2018                         13886     4153.733978   
ago                Angola  2018                          5725     3289.646664   
arg             Argentina  2018                         18282    11633.498009   

              Population  Democracy index    Regime type  \
Country code                                               
afg           37172386.0             29.7  Authoritarian   
alb            2866376.0             59.8         Hybrid   
dza           42228429.0             35.0  Authoritarian   
ago           30809762.0             36.2  Authoritarian   
arg           44494502.0             70.2      Democracy   

              Fuel exports (% of merchandise exports)     GDP total  \
Country code                                                          
afg                                         10.213180  1.835388e+10   
alb                                          1.662666  1.514702e+10   
dza                                         96.111562  1.754057e+11   
ago                                         92.417810  1.013532e+11   
arg                                          4.119375  5.176267e+11   

              GDP total (PPP dollars)  
Country code                           
afg                      6.449409e+10  
alb                      3.527362e+10  
dza                      5.863840e+11  
ago                      1.763859e+11  
arg                      8.134485e+11

### Save data frame as CSV for further usage

Finally, we save the cleaned data set into a CSV file.

In [53]:
import os

if not os.path.exists(output_folder):
    os.mkdir(output_folder)

df.to_csv(f"{output_folder}/democracy_and_gdp.csv")